perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -0,0 +1,997 @@
+{
+  return -calcRightLateralOffset(boundary_line, search_pose);
+}
+
+void MapBasedPredictionNode::updateFuturePossibleLanelets(
+  const TrackedObject & object, const lanelet::routing::LaneletPaths & paths)
+{
+  std::string object_id = toHexString(object.object_id);
+  if (objects_history_.count(object_id) == 0) {
+    return;
+  }
+
+  std::vector<lanelet::ConstLanelet> & possible_lanelets =
+    objects_history_.at(object_id).back().future_possible_lanelets;
+  for (const auto & path : paths) {
+    for (const auto & lanelet : path) {
+      bool not_in_buffer = std::find(possible_lanelets.begin(), possible_lanelets.end(), lanelet) ==
+                           possible_lanelets.end();
+      if (not_in_buffer) {
+        possible_lanelets.push_back(lanelet);
+      }
+    }
+  }
+}
+
+void MapBasedPredictionNode::addReferencePaths(
+  const TrackedObject & object, const lanelet::routing::LaneletPaths & candidate_paths,
+  const float path_probability, const ManeuverProbability & maneuver_probability,
+  const Maneuver & maneuver, std::vector<PredictedRefPath> & reference_paths)
+{
+  if (!candidate_paths.empty()) {
+    updateFuturePossibleLanelets(object, candidate_paths);
+    const auto converted_paths = convertPathType(candidate_paths);
+    for (const auto & converted_path : converted_paths) {
+      PredictedRefPath predicted_path;
+      predicted_path.probability = maneuver_probability.at(maneuver) * path_probability;
+      predicted_path.path = converted_path;
+      predicted_path.maneuver = maneuver;
+      reference_paths.push_back(predicted_path);
+    }
+  }
+}
+
+ManeuverProbability MapBasedPredictionNode::calculateManeuverProbability(
+  const Maneuver & predicted_maneuver, const lanelet::routing::LaneletPaths & left_paths,
+  const lanelet::routing::LaneletPaths & right_paths,
+  const lanelet::routing::LaneletPaths & center_paths)
+{
+  float left_lane_change_probability = 0.0;
+  float right_lane_change_probability = 0.0;
+  float lane_follow_probability = 0.0;
+  if (!left_paths.empty() && predicted_maneuver == Maneuver::LEFT_LANE_CHANGE) {
+    left_lane_change_probability = 1.0;
+    right_lane_change_probability = 0.0;
+    lane_follow_probability = 0.9;
+  } else if (!right_paths.empty() && predicted_maneuver == Maneuver::RIGHT_LANE_CHANGE) {
+    left_lane_change_probability = 0.0;
+    right_lane_change_probability = 1.0;
+    lane_follow_probability = 0.9;
+  } else if (!center_paths.empty()) {
+    if (predicted_maneuver == Maneuver::LEFT_LANE_CHANGE) {
+      // If prediction says left change, but left lane is empty, assume lane follow
+      left_lane_change_probability = 0.0;
+      right_lane_change_probability = (!right_paths.empty()) ? 0.3 : 0.0;
+    } else if (predicted_maneuver == Maneuver::RIGHT_LANE_CHANGE) {
+      // If prediction says right change, but right lane is empty, assume lane follow
+      left_lane_change_probability = (!left_paths.empty()) ? 0.3 : 0.0;
```

Comment:
avoid `0.3` as well as the comment above.

Reply:
done